sbi/inference/trainers/vfpe/base_vf_inference.py

```diff
@@ -632,3 +659,798 @@ def _loss(
+                - `True` (or `1`): An edge exists from the row node to the column node.
+                - `False` (or `0`): No edge exists between these nodes.
+            proposal: The distribution that the inputs were sampled from.
+                Pass `None` if the parameters were sampled from the prior. Multi-round
+                training is not yet implemented, so anything other than `None` will
+                raise an error.
+            exclude_invalid_x: Whether invalid simulations are discarded during
+                training. For single-round training, it is fine to discard invalid
+                simulations, but for multi-round sequential (atomic) training,
+                discarding invalid simulations gives systematically wrong results. If
+                `None`, it will be `True` in the first round and `False` in later
+                rounds. Note that multi-round training is not yet implemented.
+            data_device: Where to store the data, default is on the same device where
+                the training is happening. If training a large dataset on a GPU with not
+                much VRAM can set to 'cpu' to store data on system memory instead.
+
+        Returns:
+            MaskedVectorFieldInference object
+                (returned so that this function is chainable).
+        """
+
+        inference_name = self.__class__.__name__
+        assert proposal is None, (
+            f"Multi-round {inference_name} is not yet implemented. "
+            f"Please use single-round {inference_name}."
+        )
+        current_round = 0
+
+        if exclude_invalid_x is None:
+            exclude_invalid_x = current_round == 0
+
+        if data_device is None:
+            data_device = self._device
+
+        batch_dims = inputs.shape[:-2]
+        num_nodes = inputs.shape[-2]
+
+        if condition_masks is None:
+            # Generate masks with Bernoulli.
+            condition_masks = torch.bernoulli(
+                torch.full((*batch_dims, num_nodes), 0.5, device=inputs.device)
+            ).bool()
+
+            # Find rows that are all True or all False.
+            all_same = condition_masks.all(dim=-1) | (~condition_masks.any(dim=-1))
+
+            # If there are any such rows, flip a random element to ensure
+            # there's at least one True and one False.
+            if all_same.any():
+                invalid_indices = torch.where(all_same)
+
+                # For each invalid row, select a random column to flip.
+                cols_to_flip = torch.randint(
+                    num_nodes, (invalid_indices[0].shape[0],), device=inputs.device
+                )
+
+                # Create full indices for flipping and apply the flip.
+                indices_to_flip = invalid_indices + (cols_to_flip,)
+                condition_masks[indices_to_flip] = ~condition_masks[indices_to_flip]
+
+        if edge_masks is None:
```

Comment:
I think most implementations of attention would work passing "None" directly. This would save some memory in this case.